### PR TITLE
Dev backend list map

### PR DESCRIPTION
**Pull request by folkertdev**
and other stuff required for running `RustGlue.roc`. we cannot run that file yet, and I've cut some corners here. But I think this is in a good state to merge, and we can do the full implementation later.

### CODE REVIEW

File: crates/compiler/test_gen/src/gen_list.rs [hunk 1]
```diff
@@ -1328,9 +1328,30 @@ fn list_map_all_inline() {
...
+        ),
+        RocList::from_slice(&[123]),
+        RocList<i64>
+    );
+}
+
 #[test]
 #[cfg(any(feature = "gen-llvm", feature = "gen-wasm"))]
-fn list_map_closure() {
+fn list_map_closure_float() {
```
Comment (folkertdev) [inline crates/compiler/test_gen/src/gen_list.rs:1354]: this float test has some issue, the int test above succeeds. So I suspect there is some problem with how we handle float registers in the dev backend.

File: crates/compiler/test_gen/src/gen_tags.rs [hunk 2]
```diff
@@ -907,7 +907,9 @@ fn alignment_in_multi_tag_pattern_match() {
                         { bool, int }
 
                     Empty ->
-                        { bool: Bool.false, int: 0 }
+                        # dev backend codegen bug means we cannot use this inline
+                        false = Bool.false
+                        { bool: false, int: 0 }
```
Comment (folkertdev) [inline crates/compiler/test_gen/src/gen_tags.rs:912]: the problem is that we have an assignment `Bool.false = Call Bool.false`, so `Bool.false` is used as a function and as a local symbol. That is an IR problem that we need to fix in mono.

File: crates/compiler/gen_dev/src/generic64/mod.rs [hunk 3]
```diff
@@ -2709,6 +2838,62 @@ impl<
...
+
+        if size - copied >= 1 {
+            for _ in (0..(size - copied)).step_by(1) {
+                ASM::mov_reg8_base32(buf, reg, from_offset + copied);
+                ASM::mov_base32_reg8(buf, to_offset + copied, reg);
+
+                copied += 1;
+            }
+        }
+        */
```
Comment (bhansconnect) [inline crates/compiler/gen_dev/src/generic64/mod.rs:2894]: Do you still want this commented out? Any particular reason why?

File: crates/compiler/test_gen/src/gen_list.rs (hunk 1 above)
Comment (bhansconnect) [reply crates/compiler/test_gen/src/gen_list.rs:1354]: I would guess it is in function argument setting/loading, but that is a very loose guess.

File: crates/compiler/gen_dev/src/generic64/mod.rs [hunk 4]
```diff
@@ -1531,38 +1437,178 @@ impl<
...
+                let caller_string = self.function_symbol_to_string(
+                    caller_proc.proc_symbol,
+                    std::iter::empty(),
+                    None,
+                    Layout::UNIT,
+                );
+
+                self.caller_procs.push(caller_proc);
+
+                let inc_n_data = Symbol::DEV_TMP5;
```
Comment (bhansconnect) [inline crates/compiler/gen_dev/src/generic64/mod.rs:1516]: Why just start with TMP5 instead of 1?

**Review by bhansconnect — APPROVED** [review]

File: crates/compiler/gen_dev/src/generic64/mod.rs (hunk 3 above)
Comment (folkertdev) [reply crates/compiler/gen_dev/src/generic64/mod.rs:2894]: will clean up later

File: crates/compiler/gen_dev/src/generic64/mod.rs (hunk 4 above)
Comment (folkertdev) [reply crates/compiler/gen_dev/src/generic64/mod.rs:1516]: happened over time. will clean up later.

**Pull request merged**